sdv/sampling/hierarchical_sampler.py

```diff
@@ -0,0 +1,238 @@
+
+        Args:
+            child_name (str):
+                The name of the child table.
+            parent_row (pd.Series):
+                The row from the parent table to use for the child synthesizer.
+        """
+        raise NotImplementedError()
+
+    def _add_foreign_key_column(self, child_table, parent_table, child_name, parent_name):
+        """Add the foreign key that connects the child table to the parent table.
+
+        Args:
+            child_table (pd.DataFrame):
+                The table containing data sampled for the child.
+            parent_table (pd.DataFrame):
+                The table containing data sampled for the parent.
+            parent_name (str):
+                The name of the parent table.
+            child_name (str):
+                The name of the child table.
+        """
+        raise NotImplementedError()
+
+    def _get_foreign_keys(self, parent_table_name, child_table_name):
+        """Get all foreign keys for the parent table."""
+        foreign_keys = []
+        for relation in self.metadata.relationships:
+            if parent_table_name == relation['parent_table_name'] and\
+               child_table_name == relation['child_table_name']:
+                foreign_keys.append(deepcopy(relation['child_foreign_key']))
+
+        return foreign_keys
+
+    def _process_samples(self, table_name, sampled_rows):
+        """Process the ``sampled_rows`` for the given ``table_name``.
+
+        Process the raw samples and convert them to the original space by reverse transforming
+        them. Also, when there are synthesizer columns (columns used to recreate an instance
+        of a synthesizer), those will be returned together.
+
+        Args:
+            table_name (str):
+                The name of the table.
+            sampled_rows (pd.DataFrame):
+                The sampled rows from the table to process.
+        """
+        data_processor = self._table_synthesizers[table_name]._data_processor
+        sampled = data_processor.reverse_transform(sampled_rows)
+
+        synthesizer_columns = list(set(sampled_rows.columns) - set(sampled.columns))
+        if synthesizer_columns:
+            sampled = pd.concat([sampled, sampled_rows[synthesizer_columns]], axis=1)
+
+        return sampled
+
+    def _sample_rows(self, synthesizer, table_name, num_rows=None):
+        """Sample ``num_rows`` from ``synthesizer``.
+
+        Args:
+            synthesizer (copula.multivariate.base):
+                The fitted synthesizer for the table.
+            table_name (str):
+                Name of the table to sample from.
+            num_rows (int):
+                Number of rows to sample.
+
+        Returns:
+            pandas.DataFrame:
+                Sampled rows, shape (, num_rows)
+        """
+        num_rows = num_rows or synthesizer._num_rows
+        if synthesizer._model:
+            sampled_rows = synthesizer._sample(num_rows)
```

Comment:
I know this is the current implementation, but I'm wondering if constraints are working properly when they require reject sampling. I believe reject sampling is handled in `sample` which we are skipping over here. Maybe we just want to call `sample()`